oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/ServerApi.java

```diff
@@ -0,0 +1,370 @@
+
+    @Named("servers:privatenetwork:list")
+    @GET
+    @Path("/{serverId}/private_networks")
+    @Fallback(Fallbacks.NullOnNotFoundOr404.class)
+    List<ServerPrivateNetwork> listPrivateNetworks(@PathParam("serverId") String serverId);
+
+    @Named("servers:privatenetwork:create")
+    @POST
+    @Path("/{serverId}/private_networks")
+    @MapBinder(BindToJsonPayload.class)
+    Server assignPrivateNetwork(@PathParam("serverId") String serverId, @PayloadParam("id") String privateNetworkId);
+
+    @Named("servers:privatenetwork:get")
+    @GET
+    @Path("/{serverId}/private_networks/{privateNetworkId}")
+    @Fallback(Fallbacks.NullOnNotFoundOr404.class)
+    PrivateNetwork getPrivateNetwork(@PathParam("serverId") String serverId, @PathParam("privateNetworkId") String privateNetworkId);
+
+    @Named("servers:privatenetwork:delete")
+    @DELETE
+    @Path("/{serverId}/private_networks/{privateNetworkId}")
+    @Fallback(Fallbacks.NullOnNotFoundOr404.class)
+    @MapBinder(BindToJsonPayload.class)
+    Server deletePrivateNetwork(@PathParam("serverId") String serverId, @PathParam("privateNetworkId") String privateNetworkId);
+
+    @Named("servers:snapshot:list")
+    @GET
+    @Path("/{serverId}/snapshots")
+    @ResponseParser(ServerApi.SnapshotListParser.class)
+    @Fallback(Fallbacks.EmptyListOnNotFoundOr404.class)
+    List<Snapshot> listSnapshots(@PathParam("serverId") String serverId);
+
+    @Named("servers:snapshot:create")
+    @POST
+    @Path("/{serverId}/snapshots")
+    Server createSnapshot(@PathParam("serverId") String serverId);
+
+    @Named("servers:snapshot:update")
+    @PUT
+    @Path("/{serverId}/snapshots/{snapshotId}")
+    @MapBinder(BindToJsonPayload.class)
+    Server restoreSnapshot(@PathParam("serverId") String serverId, @PathParam("snapshotId") String snapshotId);
+
+    @Named("servers:snapshot:delete")
+    @DELETE
+    @Path("/{serverId}/snapshots/{snapshotId}")
+    @Fallback(Fallbacks.NullOnNotFoundOr404.class)
+    @MapBinder(BindToJsonPayload.class)
+    Server deleteSnapshot(@PathParam("serverId") String serverId, @PathParam("snapshotId") String snapshotId);
+
+    @Named("servers:clone:create")
+    @POST
+    @Path("/{serverId}/clone")
+    Server clone(@PathParam("serverId") String serverId, @BinderParam(BindToJsonPayload.class) Server.Clone clone);
+
+    static final class FirewallPolicyListParser extends ParseJson<List<ServerFirewallPolicy>> {
+
+        static final TypeLiteral<List<ServerFirewallPolicy>> list = new TypeLiteral<List<ServerFirewallPolicy>>() {
+        };
+
+        @Inject
+        FirewallPolicyListParser(Json json) {
+            super(json, list);
+        }
+
+        @Override
+        public <V> V apply(InputStream stream, Type type) throws IOException {
+            try {
+                Gson gson = new GsonBuilder().registerTypeAdapterFactory(new ArrayAdapterFactory()).create();
```

Comment:
Since the ArrayAdapterFactory is now only used for these two concrete lists, would it make sense to just have two adapters, one specific for each list and remove the generic logic?